libnest/src/package/mod.rs

```diff
@@ -54,72 +60,64 @@ impl Default for Kind {
 #[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default, Serialize, Deserialize)]
 pub struct PackageName(String);
 
-impl PackageName {
-    /// Create a [`PackageName`] from a String
-    pub fn from(name: String) -> Self {
-        PackageName(name)
-    }
-}
+macro_rules! strong_name_impl {
+    ($NameType:ident, $RegexValidator:ident) => {
+        impl $NameType {
+            /// Create a $NameType from a String
+            #[inline]
+            pub fn from(name: String) -> Self {
+                $NameType(name)
+            }
+        }
 
-impl std::ops::Deref for PackageName {
-    type Target = String;
+        impl std::convert::TryFrom<&str> for $NameType {
+            type Error = Error;
+
+            #[inline]
+            fn try_from(value: &str) -> Result<Self, Self::Error> {
+                if $RegexValidator.is_match(value) {
+                    Ok(Self::from(String::from(value)))
+                } else {
+                    Err(format_err!("'{}' is not a valid name", value))
+                }
+            }
+        }
 
-    #[inline]
-    fn deref(&self) -> &String {
-        &self.0
-    }
-}
+        impl std::ops::Deref for $NameType {
+            type Target = String;
 
-impl std::convert::AsRef<Path> for PackageName {
-    #[inline]
-    fn as_ref(&self) -> &std::path::Path {
-        self.0.as_ref()
-    }
-}
+            #[inline]
+            fn deref(&self) -> &String {
```

Comment:
`&String`? Why? (I guess, why not, there's no real problem to that, but people tend to return an `&str` even when returning an `&String` was possible)